components/webext-storage/src/db.rs

```diff
@@ -0,0 +1,270 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use crate::error::*;
+use crate::schema;
+use rusqlite::types::{FromSql, ToSql};
+use rusqlite::Connection;
+use rusqlite::OpenFlags;
+use sql_support::ConnExt;
+use std::fs;
+use std::path::{Path, PathBuf};
+use std::sync::{Arc, Mutex};
+use url::Url;
+
+// These can't be consts as the `|` is actually a fn.
+fn get_write_flags() -> OpenFlags {
+    OpenFlags::SQLITE_OPEN_NO_MUTEX
+        | OpenFlags::SQLITE_OPEN_URI
+        | OpenFlags::SQLITE_OPEN_CREATE
+        | OpenFlags::SQLITE_OPEN_READ_WRITE
+}
+
+fn get_read_flags() -> OpenFlags {
+    OpenFlags::SQLITE_OPEN_NO_MUTEX | OpenFlags::SQLITE_OPEN_URI | OpenFlags::SQLITE_OPEN_READ_ONLY
+}
+
+/// The entry-point to getting a database connection. No enforcement of this
+/// as a singleton is made - that's up to the caller. If you make multiple
+/// StorageDbs pointing at the same physical database, you are going to have a
+/// bad time.
+pub struct StorageDb {
+    db_path: PathBuf,
+    pub writer: Arc<Mutex<Connection>>,
+}
+impl StorageDb {
+    /// Create a new, or fetch an already open, StorageDb backed by a file on disk.
+    pub fn new(db_path: impl AsRef<Path>) -> Result<Arc<Self>> {
+        let db_path = normalize_path(db_path)?;
+        Self::new_or_existing(db_path)
+    }
+
+    /// Create a new, or fetch an already open, memory-based StorageDb. You must
+    /// provide a name, but you are still able to have a single writer and many
+    ///  reader connections to the same memory DB open.
+    pub fn new_memory(db_path: &str) -> Result<Arc<Self>> {
+        let name = PathBuf::from(format!("file:{}?mode=memory&cache=shared", db_path));
+        Self::new_or_existing(name)
+    }
+
+    fn new_or_existing(db_path: PathBuf) -> Result<Arc<Self>> {
```

Comment:
It's not clear to me what the "or existing" part of this name is meant to indicate.

Reply:
It no longer does the 'or_existing' part :) I culled a lot of this file in the new version.